hexrd/xrdutil/nf-hedm/nf-HEDM_test.py

```diff
@@ -0,0 +1,1132 @@
+
+
+def worker_init(id_state, id_exp):
+    """process initialization function. This function is only used when the
+    child processes are spawned (instead of forked). When using the fork model
+    of multiprocessing the data is just inherited in process memory."""
+    import joblib
+
+    global _mp_state
+    state = joblib.load(id_state)
+    experiment = joblib.load(id_exp)
+    _mp_state = state + (experiment,)
+
+
+@contextlib.contextmanager
+def grand_loop_pool(ncpus, state):
+    """function that handles the initialization of multiprocessing. It handles
+    properly the use of spawned vs forked multiprocessing. The multiprocessing
+    can be either 'fork' or 'spawn', with 'spawn' being required in non-fork
+    platforms (like Windows) and 'fork' being preferred on fork platforms due
+    to its efficiency.
+    """
+    # state = ( chunk_size,
+    #           image_stack,
+    #           angles,
+    #           precomp,
+    #           coords,
+    #           experiment )
+    global _multiprocessing_start_method
+    if _multiprocessing_start_method == 'fork':
+        # Use FORK multiprocessing.
+
+        # All read-only data can be inherited in the process. So we "pass" it
+        # as a global that the child process will be able to see. At the end of
+        # theprocessing the global is removed.
+        global _mp_state
+        _mp_state = state
+        pool = multiprocessing.Pool(ncpus)
+        yield pool
+        del (_mp_state)
+    else:
+        # Use SPAWN multiprocessing.
+
+        # As we can not inherit process data, all the required data is
+        # serialized into a temporary directory using joblib. The
+        # multiprocessing pool will have the "worker_init" as initialization
+        # function that takes the key for the serialized data, which will be
+        # used to load the parameter memory into the spawn process (also using
+        # joblib). In theory, joblib uses memmap for arrays if they are not
+        # compressed, so no compression is used for the bigger arrays.
+        import joblib
+        tmp_dir = tempfile.mkdtemp(suffix='-nf-grand-loop')
+        try:
+            # dumb dumping doesn't seem to work very well.. do something ad-hoc
+            logging.info('Using "%s" as temporary directory.', tmp_dir)
+
+            id_exp = joblib.dump(state[-1],
+                                 os.path.join(tmp_dir,
+                                              'grand-loop-experiment.gz'),
+                                 compress=True)
+            id_state = joblib.dump(state[:-1],
+                                   os.path.join(tmp_dir, 'grand-loop-data'))
+            pool = multiprocessing.Pool(ncpus, worker_init,
+                                        (id_state[0], id_exp[0]))
+            yield pool
+        finally:
+            logging.info('Deleting "%s".', tmp_dir)
+            shutil.rmtree(tmp_dir)
```

Comment:
@joelvbernier It looks to me that starting in python 3.8, they deemed the fork method unsafe on mac. It now defaults to using spawn, which might be why only `--force-spawn-multiprocessing` works for you. You may have to call [`multiprocessing.set_start_method('fork')`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.set_start_method) to get Mac to fork, but even then, since it is considered unsafe, it might not work.

From [here](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods): "Changed in version 3.8: On macOS, the spawn start method is now the default. The fork start method should be considered unsafe as it can lead to crashes of the subprocess."

Reply:
I am guess at the end of the day if we are targeting HPC, Linux will be the target platform.